### PR TITLE
optimize+bugfix: improved heuristics for the PCRE MAP_JIT workaround + support for binaries without PCRE

### PR DESCRIPTION
**Pull request by thibaultcha**
> I hereby granted the copyright of the changes in this pull request
to the authors of this lua-resty-core project.

This is a follow up to d046562 which improves the heuristics to decide
whether PCRE JIT compilation should be disabled in init phase under
macOS or not. The heuristics now are:

- If under macOS
- If using PCRE 8.43 or greater
- Then PCRE JIT compilation will be disabled in the init phase

While PCRE JIT compilation is disabled as such, a notice log will notify
users attempting to use the `j` or `o` flags.

Once past the init phase, we revert the workaround and allow for the `j`
and `o` flags to be used, without any branching or additional
performance penalty.

New tests and minor refactorings have been included as well.

---

Also included: `bugfix: ensured resty.core can be loaded in binaries without PCRE support.`

### CODE REVIEW

File: lib/resty/core/regex.lua [hunk 1]
```diff
@@ -371,6 +390,79 @@ local function parse_regex_opts(opts)
 end
 
 
+if no_jit_in_init then
+    local parse_regex_opts_ = parse_regex_opts
+
+    parse_regex_opts = function(opts)
```
Comment (agentzh) [inline lib/resty/core/regex.lua]: Style: needs a space before `(`. Please address other similar places.

File: lib/resty/core/request.lua [hunk 2]
```diff
@@ -68,7 +68,7 @@ local truncated = ffi.new("int[1]")
 
 local req_headers_mt = {
     __index = function (tb, key)
-        return rawget(tb, (gsub(lower(key), '_', '-', "jo")))
+        return rawget(tb, (re.gsub(lower(key), '_', '-', "jo")))
```
Comment (agentzh) [inline lib/resty/core/request.lua]: Alas. This is slower on all architectures. Here it relies on PCRE. Maybe we should introduce a C function to do this simple character replacement?

Comment (thibaultcha) [reply lib/resty/core/request.lua]: @agentzh I just pushed a commit proposing an implementation. Do you think this new `str_replace_char` function should live in a different Lua module? `misc.lua` or a new `string.lua`? It isn't exposed as it stands and thus does not have tests yet. I will add them based on your feedback.

File: .travis.yml [hunk 3]
```diff
@@ -63,7 +63,7 @@ install:
   - git clone https://github.com/openresty/openresty.git ../openresty
   - git clone https://github.com/openresty/openresty-devel-utils.git
   - git clone https://github.com/simpl/ngx_devel_kit.git ../ndk-nginx-module
-  - git clone https://github.com/openresty/lua-nginx-module.git ../lua-nginx-module
+  - git clone -b feat/ngx_http_lua_ffi_str_replace_char https://github.com/thibaultcha/lua-nginx-module.git ../lua-nginx-module
```
Comment (thibaultcha) [inline .travis.yml]: Direct link to the FFI API on my fork: https://github.com/thibaultcha/lua-nginx-module/commit/51505338fbb5df261df670a75a2c4aab748e6b3c

File: lib/resty/core/request.lua (hunk 2 above)
Comment (agentzh) [reply lib/resty/core/request.lua]: @thibaultcha Could be in ` resty.core.base`.

Comment (agentzh) [reply lib/resty/core/request.lua]: @thibaultcha Or `resty.core.util`. Maybe this is better.

**Comment by thibaultcha** [comment]
Sister PR at https://github.com/openresty/lua-nginx-module/pull/1523 for the `str_replace_char()` FFI API.

File: lib/resty/core/regex.lua [hunk 4]
```diff
@@ -79,24 +103,36 @@ local ngx_lua_ffi_init_script_engine
...
+-- cached when running under macOS, even if the user requests so. Caching is
+-- thus disabled to prevent further calls of same regex in worker to have poor
+-- performance.
+--
+-- TODO: improve this workaround when PCRE allows for unspecifying the MAP_JIT
+-- option.
+local no_jit_in_init
 
+if jit.os == "OSX" then
+    pcre_ver = ffi_string(pcre_ver)
```
Comment (agentzh) [inline lib/resty/core/regex.lua]: It's not a great idea to use the same Lua variable name for values of completely differently types.

Comment (thibaultcha) [reply lib/resty/core/regex.lua]: I have to say I concur.

**Pull request merged**